### ISSUE TITLE
Export "folder name" from multi folder searches required

### ISSUE THREAD

**Issue by philbusby** [thundernest/import-export-tools-ng#349]
This is an excellent tool. Thank you.

An export of a **search** of multiple folders works much like an export of a normal email folder. However, a search of multiple folders includes a **folder column** in the results, which is not exported by any of the export-tools-ng options.

For example, a search of two email folders will list all emails in those two folders that match the search terms and the results include which folder the email was located. However, exporting this data only shows the normal from, to, subject, body and no the folder name. 

I would need to export each folder to csv, and manually add a column in the spreadsheet for the folder name and repeat. The job that I have contains 500 folders.

Is it possible for me to adjust the code of my import-export-tools-ng installation?

Many thanks,

**Comment by cleidigh** [comment]
@philbusby 
If I understand correctly what you are seeking is an additional column in the index file?
@cleidigh

**Comment by cleidigh** [comment]
@philbusby 
Still some clarification : 
There are two outputs where the folder name could be added, to the html index or the csv output. Reading your OP I think you want the csv addition.
If that is correct let me make an assessment.  Of course as open source you can always modif for your use, but that would probably require more effort without all the addon background.
@cleidigh

**Comment by cleidigh** [comment]
@philbusby 
So here is the output adding the full msg URL.
This is a header field I have easily.
The URL has the type, full account /server, folder and message # 
Since it would be possible for multi account searches to be performed, this fully qualifies the message. You can parse whatever in the spreadsheet.
Let me know what you think then we can refine.
@cleidigh 

```
"imap-message://test1%40kokkini.net@imap.kokkini.net/INBOX/Sent#1784,test","Christopher Leidigh <t1@kok.net>","t1@kok.net",5/04/2020, ,
"imap-message://test1%40kokkini.net@imap.kokkini.net/INBOX/TopFolder#1,test","Christopher Leidigh <t1@kok.net>","t1@kok.net",10/30/2019, ,
"imap-message://test1%40kokkini.net@imap.kokkini.net/INBOX/TopFolder/Subfolder1#1,test","Christopher Leidigh <t1@kok.net>","t1@kok.net",10/23/2019, ,
"imap-message://test1%40kokkini.net@imap.kokkini.net/INBOX/Archives/2017#1,Msg 0a5d8j6g5i","test@example.com","test@example.com",12/30/2017, ,
"imap-message://test1%40kokkini.net@imap.kokkini.net/INBOX/Sent#1780,itest","Christopher Test <t1@kok.net>","ibmsadmin1@exy.com",10/07/2014, ,
"imap-message://test1%40kokkini.net@imap.kokkini.net/INBOX/Sent#1778,file test 2","Christopher Test <t1@kok.net>","ibms@exy.com",10/06/2014, ,
"imap-message://test1%40kokkini.net@imap.kokkini.net/INBOX/Sent#1777,sending large files","Christopher Test <t1@kok.net>","ibms@exy.com",10/06/2014, ,
"imap-message://test1%40kokkini.net@imap.kokkini.net/INBOX/Sent#1776,att test","Christopher Test <t1@kok.net>","xyz@gmail.com",10/06/2014, ,
"imap-message://test1%40kokkini.net@imap.kokkini.net/INBOX/Sent#785,testt","Christopher Test <t1@kok.net>","xyz@exy.com",10/04/2014, ,

```

**Comment by cleidigh** [comment]
@philbusby 
Did you see my example above? Are you still interested in this?
@cleidigh

**Comment by cleidigh** [comment]
@philbusby 
The example above is a csv export with a new field that includes both the account and folder. As you can see their are different folders because I exported from a search 
I thought that is exactly what you were looking for?
@cleidigh

**Comment by philbusby** [comment]
You seem to be only exporting **folder names**, whereas I am exporting messages from search folders. Perhaps you misinterpreted my phrase "An export of a search of multiple folders". What I meant was a **search** is made up of emails that are in **multiple folders** (which are selected in the search configuration). This is why I want to **export the location field**, as it is the only way that I know what folder the email came from. This is nothing about exporting **multiple folder names**.

When I right click a Thunderbird _email_ folder and click importexporttoolsng > Export All messages in a folder > spreadsheet (csv) I get a messages.csv file containing all the dates, subject, body data, but no location.

When I right click a Thunderbird _search_ folder and click importexporttoolsng > Export All messages in a folder > spreadsheet (csv) I get a messages.csv file containing all the dates, subject, body data, but no location.

When I export to a new mbox, rather than a spreadsheet (csv), the new folder contains emails that have dates, from, to, subject, body, but the location column is empty.

To replicate this; create a search of a common word and select multiple folders. Then try to export these email messages to a csv or new mbox. Then look at the loation field to see where the emails came from.

The importexporttoolsng export facility only exports date, from, to , subject, body, **but it does not export the location field.**

I do apologise, but I do not think I can explain it any clearer. Thanks for persevering with me.

**Comment by cleidigh** [comment]
@philbusby 
Yes I think we are still not completely understanding each other, but I think we are close.
I will describe what I believe you want AND what I have done experimentally. I will only comment on the csv as you only now mentioned mbox which is a completely different animal.

- You need to use export to csv, comma separated values, text encoded
- The current fields do not include the source folder for the message
- You would like the csv to include this new field 
- I created a test change to do this, however I made the field also include the account 
- The resulting csv output is shown above .
- That allows you to export to csv from a search.

If this still misses the point then give me an example output for the csv.
@cleidigh 

-

**Comment by philbusby** [comment]
OK, yes. You seem to have described exactly what I am looking for. I appreciate that csv/mbox exports are very different animals to the programmer, but to the layman these are just options on the import-export-tools-ng menu. I export to csv to create the WordCloud and I export to mbox which would be linked from the WordCloud.

Mea culpa! Was your csv output above listing messages? if the #_number_ after the folder name meant **message number** and not part of the folder name, then this is indeed exactly what I want.  

So, if test1%40kokkini.net@imap.kokkini.net/INBOX/Sent#1777 and test1%40kokkini.net@imap.kokkini.net/INBOX/Sent#1778 are different messages #1777 and #1778 in the Sent folder, then this is indeed exactly what I want. 

If this is available in the 12.0.2 b7 beta, would you be happy with me testing it?

Do you have any idea how incredibly good and useful your program is?

**Comment by cleidigh** [comment]
@philbusby 
Yes please give b7 a go And confirm it meets your goal.
For a release I will add a preference that can toggle the additional column. For now this would not be in the UI it would be set with the preference editor.
I appreciate your comments on IETNG. I am not the original author, however, I took it over around four years ago specifically because I was a user and as the #1 addon it was critical for the community.  I have added a lot, but I wish I could do more, but I have limits due to my disability (programming with eye gaze). 
@cleidigh

**Comment by philbusby** [comment]
b7 provides exactly what I need for exporting the folder location to "spreadsheet (csv)". 

The export to mbox still does not export the folder name and the folder column in the mbox emails is empty. This would be useful, but **not essential**, as B7 has provided the folder information in the index export.

It would be **useful** to have the mbox exporting the folder name too though! I do see the complication in having an email showing a different folder name in the email's "location" column to the current email's location. However the location is currently blank in these mbox exports, so it might be possible to achieve if we knew what email header would automatically populate this field in thunderbird read of a mbox.

Many thanks!

**Comment by cleidigh** [comment]
@philbusby 
b8 is my last beta. I will release (into review Q) in the next few days.
As mentioned, you will now need to enable the folder location column by setting the advanced preference : 
extensions.importexporttoolsng.experimental.csv.account_folder_col
to true.

https://github.com/thundernest/import-export-tools-ng/blob/v12.0.2/xpi/import-export-tools-ng-12.0.2-b8-tb.xpi

@cleidigh

**Comment by cleidigh** [comment]
Addressed with v12.0.2 Release
